gpt-auto-generator: rework/simplify logic for picking /efi or /boot

keszybz opened the pull request:
I started looking into https://github.com/uapi-group/specifications/issues/35.

BLS says:
> Otherwise [no existing XBOOTLDR partition], if on GPT and an ESP is found and
> it is large enough (let’s say at least 1G) it should be used as $BOOT and
> used as primary location to place boot loader menu resources in.

> It is recommended to mount $BOOT to /boot/, and the ESP to /efi/.

DPS says:
> The ESP used for the current boot is automatically mounted to /efi/ (or
> /boot/ as fallback), unless a different partition is mounted there (possibly
> via /etc/fstab, or because the Extended Boot Loader Partition — see below —
> exists) or the directory is non-empty on the root disk.

I don't think we want to mount the same partition in two places. If the same partition is not mounted in two places, then the two specs are contradictory.

The code in gpt-auto-generator implemented the logic from the DPS. It is modified to implement the logic from BLS.

Effectively:
- if both /boot and /efi exist:
  - if both XBOOTLDR and ESP exist: ESP on /efi, XBOOTLDR on /boot
  - if only ESP exists: ESP on /boot
  - if only XBOOTLDR exists: XBOOTLDR on /boot
- if only /boot exists:
  - if XBOOTLDR exists: XBOOTLDR on /boot
  - if only ESP exists: ESP on /boot
- if only /efi exists:
  - if ESP exists: ESP on /efi Not mounting stuff because the mount points don't exist is treated as success.

Thus, the generator attempts to match the partitions and mount points to the extent possible. In all cases, /boot is the primary place to install kernels. ESP can be found on /boot or /efi, depending on the situation.

If this patch is merged, I'll submit fixes for BLS and DPS to describe the same logic.

YHNdnzj commented:
Hmm, I'd like to add one more case, that both `/boot` and `/efi` exist but the former is not empty. In this case, we should also mount the ESP to `/efi` so that user still gets the auto-discovered ESP mounted.

keszybz commented:
> Hmm, I'd like to add one more case, that both `/boot` and `/efi` exist but the former is not empty, and there's no XBOOTLDR. In this case, we should also mount the ESP to `/efi` so that user still gets the auto-discovered ESP mounted.

That's a good point. Amended like this.

keszybz commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -526,42 +570,48 @@ static int add_partition_esp(DissectedPartition *p, bool has_xbootldr) {
+                        log_warning_errno(errno,
+                                          "/efi does not exist, no place to mount ESP configured or detected.");
+                        return 0;
+                } else {
```
Indeed. And we need to check if `esp_path` was set once again, because it is unconditionally used later on. Updated.

yuwata reviewed (commented):
(Not reviewed in detail, but please split this into at least two: cleanup commit and rework or fix.)

poettering commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -467,6 +468,45 @@ static int add_automount(
+        return r;
+}
+
+static int slash_efi_in_fstab(void) {
```
ditto

poettering commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -467,6 +468,45 @@ static int add_automount(
+}
+
+static bool slash_boot_exists(void) {
+        static int r = -1;
```
ditto

keszybz replied:
I changed this one to use `cache`.

poettering commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -500,17 +543,18 @@ static int add_partition_xbootldr(DissectedPartition *p) {
-                             options,
-                             "Boot Loader Partition",
-                             120 * USEC_PER_SEC);
+                return log_error_errno(r, "Failed to determine default mount options for /boot: %m");
```
/boot → /boot/?

poettering commented:
does this match what dissect-image.c would do btw?

keszybz commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -467,6 +468,45 @@ static int add_automount(
 }
 
+static int slash_boot_in_fstab(void) {
+        static int r = -1;
```
There is no need for a second variable. So we'd end up with stuff like
```C
        cache = fstab_is_mount_point("/boot");
        if (cache < 0)
                return log_error_errno(cache, "Failed to parse fstab: %m");
```
or an additional copying step.

I can switch to this if you insist, but it doesn't seem better.

poettering replied:
i'd prefer if the variable was named "cache", not "r", if it's static. that's all.

no need for two vars

keszybz replied:
OK.

keszybz commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -478,14 +518,17 @@ static int add_partition_xbootldr(DissectedPartition *p) {
         }
 
+        if (!slash_boot_exists())
+                return 0;
```
Did the previous version of the code create the directory? I don't see any provision to create the mount point… I can drop the check obviously, but wouldn't that just mean that the unit would fail if started?

keszybz replied:
Oh, it indeed does create the directory. I didn't know that. I'll rework the patch to ignore a missing directory.

poettering replied:
the dir is created by the .mount unit, if needed.

this is the only thing that works, because the fs we need to create the dir in might not be writable any earlier then immediately before creating the mount.

keszybz replied:
Yeah. Also the automount unit.

keszybz commented:
> Not reviewed in detail, but please split this into at least two: cleanup commit and rework or fix.

There is no clear split. It's changing the logic a bit and reworking the code to remove duplication because of the change. (I mean: it'd be possible to split some of the trivial changes out, but I think it'd look somewhat silly.)

Updated:
- I didn't know that the mount point directory will be created by the automount unit. I adjusted the code to treat a missing directory as "available".
- I updated the logic along the same lines, see the commit message.

I didn't rename `r` in two of the three places, as discussed above.

poettering commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -526,43 +567,46 @@ static int add_partition_esp(DissectedPartition *p, bool has_xbootldr) {
+        if (!esp_path) {
+                r = RET_NERRNO(access("/efi", F_OK));
+                if (r < 0 && r != -ENOENT)
+                        return log_error_errno(r, "Failed to determine whether /efi/ exists: %m");
```
you are not actually using the result of this test?

keszybz replied:
It checks that the dir is accessible. But I guess we don't need to do that. I'll drop it.

poettering commented on src/shared/dissect-image.c:
```diff
@@ -2110,45 +2123,36 @@ int dissected_image_mount(
-        xbootldr_mounted = mount_partition(PARTITION_XBOOTLDR, m->partitions + PARTITION_XBOOTLDR, where, "/boot", uid_shift, uid_range, flags);
-        if (xbootldr_mounted < 0)
-                return xbootldr_mounted;
+        int slash_boot_is_available = -ENODATA;
```
if you initialize this in the very next line i'd just drop the initialization here

poettering commented on src/shared/dissect-image.c:
```diff
@@ -2019,14 +2019,27 @@ static int mount_root_tmpfs(const char *where, uid_t uid_shift, DissectImageFlag
+        if (r < 0)
+                return r;
+
+        return dir_is_empty(p, /* ignore_hidden_or_backup= */ false);
```
probably should convert ENOTDIR into "false"

poettering commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -279,12 +278,14 @@ static int add_mount(
+        /* The directory will be created by the mount or automount unit when it is started. */
         if (r == -ENOENT)
                 return false;
 
```
and here too, btw, looking at this, i think we should convert ENOTDIR from the dir_is_empty() call to false

poettering commented on src/shared/dissect-image.c:
```diff
@@ -2110,45 +2123,36 @@ int dissected_image_mount(
+                r = mount_partition(PARTITION_XBOOTLDR, m->partitions + PARTITION_XBOOTLDR, where, "/boot", uid_shift, uid_range, flags);
+                if (r < 0)
+                        return r;
+                slash_boot_is_available = r;
```
hmm, shouldn't this be `!r`? after all if you just mounted XBOOTLDR to /boot then /boot is not available for anything esle.

keszybz replied:
Arrgh. Yes.

YHNdnzj commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -293,13 +293,17 @@ static int path_is_busy(const char *where) {
+        if (r == -ENOTDIR) {
+                log_debug("\"%s\" is not a directory, ignoring.", where);
+                return true;
+        } else if (r < 0)
```
Unneeded `else`

keszybz replied:
It's there on purpose to avoid extra lines. The whole thing is more readable with each indented block preceded by exactly one line with a conditional.

YHNdnzj commented on src/gpt-auto-generator/gpt-auto-generator.c:
```diff
@@ -293,13 +293,17 @@ static int path_is_busy(const char *where) {
                 return log_warning_errno(r, "Cannot check if \"%s\" is empty: %m", where);
-        if (r > 0)
-                return false;
+        else if (r == 0) {
```
same

keszybz commented:
centos ci (arch linux + sanitizers) fails with a hang/timeout in test-29-portable. I think this is https://github.com/systemd/systemd/issues/24147:
```
May 26 16:19:21 H systemd-udevd[361]: Cleanup idle workers
May 26 16:19:43 H systemd[1]: Got message type=signal sender=org.freedesktop.DBus destination=n/a path=/org/freedesktop/DBus interface>
May 26 16:19:43 H systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus int>
May 26 16:19:43 H systemd[1]: Received SIGCHLD from PID 420 (bash).
May 26 16:19:43 H systemd[1]: Child 420 (bash) died (code=exited, status=124/n/a)
May 26 16:19:43 H systemd[1]: testsuite-29.service: Child 420 belongs to testsuite-29.service.
May 26 16:19:43 H systemd[1]: testsuite-29.service: Main process exited, code=exited, status=124/n/a
May 26 16:19:43 H systemd[1]: testsuite-29.service: Failed with result 'exit-code'.
```

But test-58-repart was failing for a good reason. I reworked the second patch to actually use `/efi` if `/boot` does not exist. This matches more closely what the first patch does, and what the code was doing previously.
